File: CS/T415077/UIEditors/FilteredFileNameEditor.cs
using System;
using System.ComponentModel;
using System.Drawing.Design;
using System.Windows.Forms;

namespace T415077.UIEditors {
    internal class FilteredFileNameEditor : UITypeEditor {
        private OpenFileDialog ofd = new OpenFileDialog();
        public override UITypeEditorEditStyle GetEditStyle(
         ITypeDescriptorContext context) {
            return UITypeEditorEditStyle.Modal;
        }
        public override object EditValue(
         ITypeDescriptorContext context,
         IServiceProvider provider,
         object value) {
            ofd.FileName = value.ToString();
            ofd.Filter = "Text File|*.txt|All Files|*.*";
            if (ofd.ShowDialog() == DialogResult.OK) {
                return ofd.FileName;
            }
            return base.EditValue(context, provider, value);
        }
    }
}

File: VB/T415077/UIEditors/FilteredFileNameEditor.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing.Design
Imports System.Windows.Forms

Namespace T415077.UIEditors

    Friend Class FilteredFileNameEditor
        Inherits UITypeEditor

        Private ofd As OpenFileDialog = New OpenFileDialog()

        Public Overrides Function GetEditStyle(ByVal context As ITypeDescriptorContext) As UITypeEditorEditStyle
            Return UITypeEditorEditStyle.Modal
        End Function

        Public Overrides Function EditValue(ByVal context As ITypeDescriptorContext, ByVal provider As IServiceProvider, ByVal value As Object) As Object
            ofd.FileName = value.ToString()
            ofd.Filter = "Text File|*.txt|All Files|*.*"
            If ofd.ShowDialog() = DialogResult.OK Then
                Return ofd.FileName
            End If

            Return MyBase.EditValue(context, provider, value)
        End Function
    End Class
End Namespace
